cura/Machines/MachineNode.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) 2019 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+
+from typing import Dict, List
+
+from UM.Logger import Logger
+from UM.Signal import Signal
+from UM.Util import parseBool
+from UM.Settings.ContainerRegistry import ContainerRegistry  # To find all the variants for this machine.
+from cura.Machines.ContainerNode import ContainerNode
+from cura.Machines.QualityChangesGroup import QualityChangesGroup  # To construct groups of quality changes profiles that belong together.
+from cura.Machines.QualityGroup import QualityGroup  # To construct groups of quality profiles that belong together.
+from cura.Machines.QualityNode import QualityNode
+from cura.Machines.VariantNode import VariantNode
+
+##  This class represents a machine in the container tree.
+#
+#   The subnodes of these nodes are variants.
+class MachineNode(ContainerNode):
+    def __init__(self, container_id: str) -> None:
+        super().__init__(container_id)
+        self.variants = {}  # type: Dict[str, VariantNode] # Mapping variant names to their nodes.
+        self.global_qualities = {}  # type: Dict[str, QualityNode] # Mapping quality types to the global quality for those types.
+        self.materialsChanged = Signal()  # Emitted when one of the materials underneath this machine has been changed.
+
+        container_registry = ContainerRegistry.getInstance()
+        try:
+            my_metadata = container_registry.findContainersMetadata(id = container_id)[0]
+        except IndexError:
+            Logger.log("Unable to find metadata for container %s", container_id)
+            my_metadata = {}
+        # Some of the metadata is cached upon construction here.
+        # ONLY DO THAT FOR METADATA THAT DOESN'T CHANGE DURING RUNTIME!
+        # Otherwise you need to keep it up-to-date during runtime.
+        self.has_materials = parseBool(my_metadata.get("has_materials", "true"))
```

Comment:
Similar thing with `has_materials` to `has_variants` (see comments below) that it can be changed in the `GlobalStack`'s metadata. See:

https://github.com/Ultimaker/Cura/blob/bc4809d8f87022bf670da52c05449fd116c8cd08/plugins/MachineSettingsAction/MachineSettingsAction.py#L88-L112

Reply:
Fixed this [here](https://github.com/Ultimaker/Cura/pull/6269/commits/977beb8dbb0aa92c7dc450a2cbfbbb65ab53faf9) and fixed a subsequent bug [here](https://github.com/Ultimaker/Cura/pull/6269/commits/eaf649023a1e73533fc49a162a7af513c11ec603).